### PR TITLE
82 point pre annotating

### PR DESCRIPTION
**Pull request by memeramita**
<h2>Context</h2>
#82 We want to add point annotation as an option when the user creates an annotation template. By doing so, the default field is hidden. After the template is created, the 'Select' button is disabled before the annotation starts. This PR also makes sure that the user can save and load a template containing point annotations in the json format.

<h2>Changes</h2>

1. **`annotation_item.py`**

`__init__()`:
- Added point as an option in `self.type_selection_combo` to allow the user to select when they create a template.
- Replaced `default_label` with `self.default_label` as I need to explicitly call `.show()` and `.hide()` in order to hide it for point annotation items.

`fill_vals_point()`: for existing point annotations, set the option to "point" and fill in the annotation name

`type_changed()`:

- It checks if `default_widget` exists before removing it. We need to do this as point annotations do not have default widgets to avoid errors.
- I added an if statement that hides `default_label`, `default_options`, and `default_options_label` when point annotation is selected.

`get_data()`: Added an if statement to set `type` and `default` for point annotation.

2. **`annotation_widget.py`**

`add_existing_item()`: Added an if statement to fill in values for existing point annotations.

3. **`template_item.py`**

`ItemType`: Added `"point"` as `POINT`.

`__init__()`: If it's a point annotation item, center and disable the button. Otherwise, add the name and the widget to the layout as usual and enable the widget.

4. **`template_item.py`**

`add_item()`: added an if statement for point annotation which sets `annot_type` and creates a QPushButton widget.

5. Write tests for new and modified functions in **`annotation_item_test.py`**, **`annotation_widget_test.py`**, **`template_list_test.py`**

### CODE REVIEW

File: napari_allencell_annotator/widgets/annotation_item.py [hunk 1]
```diff
@@ -152,6 +153,18 @@ def fill_vals_list(self, name: str, default: str, options: List[str]):
...
+        Fill in name for point.
+
+        Parameters
+        ----------
+        name : str
+            a name for the annotation
+        """
+        self.type_selection_combo.setCurrentText("point")
+        self.name.setText(name)
+
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/annotation_item.py:167]: what is this function responsible for?

File: napari_allencell_annotator/widgets/annotation_item.py [hunk 2]
```diff
@@ -161,29 +174,41 @@ def _type_changed(self, text: str):
...
             self.layout.addWidget(self.default_check, 0, 7, 1, 2)
-        else:
+
+        elif text == "dropdown":
+            self.default_label.show()
             self.default_options.show()
             self.default_options_label.show()
             self.layout.addWidget(self.default_text, 0, 7, 1, 2)
 
+        else:
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/annotation_item.py]: should this be an elif for `point`?

File: napari_allencell_annotator/widgets/annotation_item.py [hunk 3]
```diff
@@ -161,29 +174,41 @@ def _type_changed(self, text: str):
         text : str
             the new type selected.
         """
-        default_widget = self.layout.itemAtPosition(0, 7).widget()
-        default_widget.setParent(None)
-        self.layout.removeWidget(default_widget)
+        default_item: QLayoutItem = self.layout.itemAtPosition(0, 7)
+        if default_item is not None:
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/annotation_item.py:178]: when would this be `None`

File: napari_allencell_annotator/widgets/annotation_widget.py [hunk 4]
```diff
@@ -52,6 +52,8 @@ def add_existing_item(self, name: str, key_info: Key | ComboKey):
             item.fill_vals_check(name, key_info.get_default_value())
         elif annot_type == "list":
             item.fill_vals_list(name, key_info.get_default_value(), key_info.get_options())
+        else:
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/annotation_widget.py]: elif

File: napari_allencell_annotator/widgets/template_item.py [hunk 5]
```diff
@@ -37,10 +39,16 @@ def __init__(self, parent: QListWidget, name: str, type: ItemType, default: Any,
...
-        self.layout.addWidget(self.editable_widget)
 
-        self.editable_widget.setEnabled(True)
+        if self.type != ItemType.POINT:
+            self.layout.addWidget(self.name)
+            self.layout.addWidget(self.editable_widget)
+            self.editable_widget.setEnabled(True)
+        else:
+            self.layout.addWidget(self.name, 0)
+            self.layout.addWidget(self.editable_widget, 1, alignment=Qt.AlignCenter)
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/template_item.py]: why are these indexed?

File: napari_allencell_annotator/widgets/template_list.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 from typing import Any, List
 
+from PyQt5.QtWidgets import QPushButton
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/template_list.py]: this need to be imported under qtpy below

File: napari_allencell_annotator/widgets/template_list.py [hunk 7]
```diff
@@ -109,6 +110,11 @@ def add_item(self, name: str, key: Key | ComboKey):
                 for opt in key.get_options():
                     widget.addItem(opt)
             widget.setCurrentText(default)
+        else:
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/template_list.py]: elif

File: napari_allencell_annotator/widgets/template_list.py [hunk 8]
```diff
@@ -109,6 +110,11 @@ def add_item(self, name: str, key: Key | ComboKey):
...
+            annot_type = ItemType.POINT
+            widget = QPushButton("Select")
+            widget.setFixedWidth(200)
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/template_list.py]: I dont think styling should go here. unless there's a good reason to do so. You can apply a style when using addwidget() instead

File: napari_allencell_annotator/_tests/widgets/annotation_widget_test.py [hunk 9]
```diff
@@ -1,9 +1,63 @@
...
+    return AnnotationWidget()
+
+
+def test_add_existing_item_string(annotation_widget: AnnotationWidget) -> None:
+    # ACT
+    annotation_widget.add_existing_item("test_string", Key("string", "test_value"))
+
+    # ASSERT
+    assert annotation_widget.count() == 1
+    assert annotation_widget.item(0).name.text() == "test_string"
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/widgets/annotation_widget_test.py]: dont need to assert item state imo

File: napari_allencell_annotator/_tests/widgets/template_list_test.py [hunk 10]
```diff
@@ -1,13 +1,124 @@
...
+
+    # ASSERT
+    assert template_list.count() == 1
+
+    item: TemplateItem = template_list.item(0)
+    assert item.name.text() == "test_string"
+    assert item.type == ItemType.STRING
+    assert item.default == "test_value"
+    assert isinstance(item.editable_widget, QLineEdit)
+    assert item.editable_widget.text() == "test_value"
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/widgets/template_list_test.py]: not needed, checked above

File: napari_allencell_annotator/_tests/widgets/template_list_test.py [hunk 11]
```diff
@@ -1,13 +1,124 @@
...
+    assert template_list.count() == 1
+
+    item: TemplateItem = template_list.item(0)
+    assert item.name.text() == "test_list"
+    assert item.type == ItemType.LIST
+    assert item.default == "a"
+    assert isinstance(item.editable_widget, QComboBox)
+    assert item.editable_widget.itemText(0) == "a"
+    assert item.editable_widget.itemText(1) == "b"
+    assert item.editable_widget.currentText() == "a"
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/widgets/template_list_test.py]: not needed, you're testing beyond TemplateList

File: napari_allencell_annotator/_tests/widgets/annotation_item_test.py [hunk 12]
```diff
@@ -52,66 +126,6 @@ def test_fill_vals_list(self):
         self._item.default_text.setText.assert_called_once_with("one")
         self._item.default_options.setText.assert_called_once_with("one, two, three")
 
-    def test_type_changed_dropdown(self):
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/widgets/annotation_item_test.py:55]: why were these deleted?

File: napari_allencell_annotator/_tests/widgets/annotation_item_test.py [hunk 13]
```diff
@@ -1,15 +1,89 @@
...
+
+
+@pytest.fixture
+def annotation_item(annotation_widget, qtbot) -> AnnotationItem:
+    return AnnotationItem(annotation_widget)
+
+
+def test_type_changed_text(annotation_item: AnnotationItem) -> None:
+    # ACT
+    annotation_item._type_changed("text")
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/widgets/annotation_item_test.py:27]: private functions do not need tests, only public ones do.

File: napari_allencell_annotator/widgets/template_item.py [hunk 14]
```diff
@@ -37,10 +39,16 @@ def __init__(self, parent: QListWidget, name: str, type: ItemType, default: Any,
...
         self.name = QLabel(name)
-        self.layout.addWidget(self.name)
-        self.layout.addWidget(self.editable_widget)
 
-        self.editable_widget.setEnabled(True)
+        if self.type != ItemType.POINT:
+            self.layout.addWidget(self.name)
+            self.layout.addWidget(self.editable_widget)
+            self.editable_widget.setEnabled(True)
+        else:
```
Comment (yrkim98) [inline napari_allencell_annotator/widgets/template_item.py:47]: elif

File: napari_allencell_annotator/widgets/annotation_item.py (hunk 2 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/annotation_item.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/6bfbf5e42452477f6bdbed152dcf5f16c8a90a38

File: napari_allencell_annotator/widgets/annotation_widget.py (hunk 4 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/annotation_widget.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/6bfbf5e42452477f6bdbed152dcf5f16c8a90a38

File: napari_allencell_annotator/widgets/template_list.py (hunk 7 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/template_list.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/6bfbf5e42452477f6bdbed152dcf5f16c8a90a38

File: napari_allencell_annotator/widgets/annotation_item.py (hunk 3 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/annotation_item.py:178]: when we change the type from point

File: napari_allencell_annotator/widgets/template_item.py (hunk 5 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/template_item.py]: they used to be stretch parameters but were removed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/f0a9cde427560fa25e875b67b0d7e3576c4edc5a

File: napari_allencell_annotator/widgets/template_list.py (hunk 8 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/template_list.py]: moved in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/f0a9cde427560fa25e875b67b0d7e3576c4edc5a

File: napari_allencell_annotator/widgets/template_list.py (hunk 6 above)
Comment (memeramita) [reply napari_allencell_annotator/widgets/template_list.py]: changed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/b7f16e45109a4a54451ebd3c5e8e8e1ae8da3f4b

File: napari_allencell_annotator/_tests/widgets/annotation_widget_test.py (hunk 9 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/widgets/annotation_widget_test.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/c7d87a520c34ddc25c82b0943d637bc2c3985e6d

File: napari_allencell_annotator/_tests/widgets/template_list_test.py (hunk 10 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/widgets/template_list_test.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/7f445d5488a221349dad061156ede36c7ed49a37

File: napari_allencell_annotator/_tests/widgets/template_list_test.py (hunk 11 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/widgets/template_list_test.py]: removed in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/7f445d5488a221349dad061156ede36c7ed49a37

File: napari_allencell_annotator/_tests/widgets/annotation_item_test.py (hunk 12 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/widgets/annotation_item_test.py:55]: I wrote new tests for these

File: napari_allencell_annotator/_tests/widgets/template_list_test.py (hunk 10 above)
Comment (yrkim98) [reply napari_allencell_annotator/_tests/widgets/template_list_test.py]: took out too many assertions. please keep all logic that is in the `add_item` function tested.

Comment (memeramita) [reply napari_allencell_annotator/_tests/widgets/template_list_test.py]: added assertions back to tests in https://github.com/aics-int/napari-allencell-annotator/pull/92/commits/6eada85860f445d23ef9230cab3a6fd349e03505

**Review by yrkim98 — APPROVED** [review]
